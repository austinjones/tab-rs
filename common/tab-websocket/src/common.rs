use crate::WebSocket;
use futures::SinkExt;
use log::error;
use serde::Serialize;
use tungstenite::{Error, Message};

pub async fn send_message<Msg: Serialize>(websocket: &mut WebSocket, message: Msg) {
    let encoded = bincode::serialize(&message);

    if let Err(e) = encoded {
        error!("failed to encode message: {}", e);
        return;
    }

    let sent = websocket.send(Message::Binary(encoded.unwrap())).await;
    if let Err(e) = sent {
        error!("failed to send message: {}", e);
    }
}

pub async fn send_close(websocket: &mut WebSocket) {
    if let Err(e) = websocket.send(Message::Close(None)).await {
        match e {
            Error::ConnectionClosed | Error::AlreadyClosed => {
                return;
            }
            _ => {
                error!("failed to send close frame: {}", e);
            }
        }
    }
}
