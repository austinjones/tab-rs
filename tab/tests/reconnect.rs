use insta::assert_snapshot;

mod common;
use common::*;

/// Tests that a session can be established, disconnected from, and re-established
/// Covers connection, ctrl-W, disconnection, and scrollback
#[tokio::test]
async fn reconnect() -> anyhow::Result<()> {
    let mut session = TestSession::new()?;

    for i in 0..10 {
        reconnect_iter(&mut session, i).await?;
    }

    Ok(())
}

async fn reconnect_iter(session: &mut TestSession, iter: usize) -> anyhow::Result<()> {
    let tab = format!("reconnect/{}/", iter);
    let result = session
        .command()
        .tab(tab.as_str())
        .await_stdout("$", 1000)
        .stdin("echo foo\n")
        .await_stdout("echo foo", 300)
        .await_stdout("$", 300)
        .stdin_bytes(&[23u8])
        .run()
        .await?;

    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("before", result.snapshot);

    let result = session
        .command()
        .tab(tab.as_str())
        .await_stdout("foo", 1000)
        .stdin("exit\n")
        .await_stdout("exit", 300)
        .complete_snapshot()
        .run()
        .await?;
    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("after", &result.snapshot);

    Ok(())
}
