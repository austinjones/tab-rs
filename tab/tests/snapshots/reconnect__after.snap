---
source: tab/tests/reconnect.rs
expression: result.stdout
---
$ echo foo
foo
$ exit
exit

